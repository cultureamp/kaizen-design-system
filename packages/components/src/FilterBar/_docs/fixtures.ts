/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { Factory, faker } from "@cultureamp/frontend-testing"
import { FilterOption } from "../types"

const userFactory = Factory.Sync.makeFactory<FilterOption>({
  value: Factory.each(i => faker.datatype.uuid()),
  label: Factory.each(
    () => `${faker.name.firstName()} ${faker.name.lastName()}`
  ),
})

const departmentFactory = Factory.Sync.makeFactory<FilterOption>({
  value: Factory.each(i => faker.datatype.uuid()),
  label: Factory.each(() => faker.commerce.department()),
})

const ascSort = (a: FilterOption, b: FilterOption) =>
  a.label.toLocaleLowerCase() > b.label.toLocaleLowerCase() ? 1 : -1

export const users = userFactory.buildList(30).sort(ascSort)
export const departments = departmentFactory.buildList(8).sort(ascSort)
