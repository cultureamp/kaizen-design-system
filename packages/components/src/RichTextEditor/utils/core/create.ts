import { EditorState, Transaction } from "prosemirror-state"
import { EditorView as ProseMirrorEditorView } from "prosemirror-view"
import { CommandOrTransaction, EditorView } from "./types"

type EditorAPI = {
  destroy: () => void
  dispatchTransaction: (maybeCommand: CommandOrTransaction) => void
}

type EditorArgs = {
  initialEditorState: EditorState
  node: HTMLElement
  onChange: (editorState: EditorState) => void
  /*
   * Pass in HTML attributes into the parent RTE node
   */
  attributes?: {
    [name: string]: string
  }
  isEditable?: () => boolean
}

/**
 * createRichTextEditor
 * Initialize a ProseMirror EditorView
 */
export function createRichTextEditor({
  initialEditorState,
  node,
  onChange = () => undefined,
  attributes,
  isEditable = () => true,
}: EditorArgs): EditorAPI {
  let editorView: EditorView | undefined

  // Handle transactions eminating from the EditorView instance
  function dispatch(tx: Transaction): void {
    const newEditorState = editorView && editorView.state.apply(tx)
    if (newEditorState) {
      onChange(newEditorState)
      editorView && editorView.updateState(newEditorState)
    }
  }

  // Allow the dispatcher to handle either a Command or a Transaction so we can
  // change the state more ergonomically upstream
  function dispatchCommandOrTransaction(
    commandOrTransaction: CommandOrTransaction
  ) {
    if (commandOrTransaction instanceof Transaction) {
      dispatch(commandOrTransaction)
    } else if (editorView) {
      commandOrTransaction(editorView.state, dispatch)
    }
  }

  editorView = new ProseMirrorEditorView(node, {
    state: initialEditorState,
    dispatchTransaction: dispatchCommandOrTransaction,
    attributes,
    editable: isEditable,
  })

  return {
    destroy: () => {
      editorView && editorView.destroy()
    },
    dispatchTransaction: dispatchCommandOrTransaction,
  }
}
