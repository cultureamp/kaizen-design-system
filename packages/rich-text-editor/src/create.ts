import { EditorState, Transaction } from "prosemirror-state"
import { EditorView } from "prosemirror-view"
// import { CAEditorView, MaybeCommand } from "./types.d"

type EditorAPI = {
  destroy: () => void
  dispatchTransaction: (maybeCommand: any) => void
}

type EditorArgs = {
  initialEditorState: EditorState
  node: HTMLElement
  onChange: (editorState: EditorState) => void
}

export function createRichTextEditor({
  initialEditorState,
  node,
  onChange = () => undefined,
}: EditorArgs): EditorAPI {
  // editorView is reassigned below ¯\_(ツ)_/¯
  // eslint-disable-next-line prefer-const
  let editorView: any

  // Handle transactions eminating from the EditorView instance
  function dispatch(tx: Transaction): void {
    const newEditorState = editorView && editorView.state.apply(tx)
    if (newEditorState) {
      onChange(newEditorState)
      editorView && editorView.updateState(newEditorState)
    }
  }

  function dispatchTransaction(maybeCommand: any) {
    if (maybeCommand instanceof Transaction) {
      dispatch(maybeCommand)
    } else if (editorView) {
      maybeCommand(editorView.state, dispatch)
    }
  }

  editorView = new EditorView(node, {
    state: initialEditorState,
    dispatchTransaction,
  })

  return {
    destroy: () => {
      editorView && editorView.destroy()
    },
    dispatchTransaction,
  }
}
