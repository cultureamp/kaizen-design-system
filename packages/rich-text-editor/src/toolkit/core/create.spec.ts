import { describe, expect, it, jest } from "@jest/globals"
import { findByText, queryByText } from "@testing-library/dom"
import { Command, EditorState, Transaction } from "prosemirror-state"
import { createRichTextEditor } from "./create"
import { testEditorState } from "./fixtures/testState"

describe("createRichTextEditor", () => {
  const attributes = { "aria-labelledby": "label-text-123" }

  it("initializes an editor with the correct content", async () => {
    const node = document.createElement("div")
    const onChange = jest.fn()

    createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
    })

    await findByText(node, "Example content")
  })

  it("returns the expected API shape", async () => {
    const node = document.createElement("div")
    const onChange = jest.fn()

    const returnValue = createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
    })

    expect(Object.keys(returnValue)).toEqual(["destroy", "dispatchTransaction"])
    expect(typeof returnValue.destroy).toEqual("function")
    expect(typeof returnValue.dispatchTransaction).toEqual("function")
  })

  it("destroys the instance", async () => {
    const node = document.createElement("div")
    const onChange = jest.fn()

    const { destroy } = createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
    })

    await findByText(node, "Example content")

    destroy()

    expect(await queryByText(node, "Example content")).toBeNull()
  })

  it("updates the DOM when commands are dispatched", async () => {
    const node = document.createElement("div")
    const onChange = jest.fn()

    const command = (
      state: EditorState,
      dispatch?: (tx: Transaction) => void
    ) => {
      // Insert text at the current selection point, which is the start because
      // we donâ€™t have a selection yet.
      if (!dispatch) return false
      dispatch(state.tr.insertText("Prepended content. "))
      return true
    }

    const { dispatchTransaction } = createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
    })

    await findByText(node, "Example content")

    dispatchTransaction(command)

    await findByText(node, "Prepended content. Example content")
  })

  it("calls onChange when the editor state changes", async () => {
    const node = document.createElement("div")
    const onChange = jest.fn()
    const command = (
      state: EditorState,
      dispatch?: (tx: Transaction) => void
    ) => {
      if (!dispatch) return false
      dispatch(state.tr.insertText("Prepended content. "))
      return true
    }

    const { dispatchTransaction } = createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
    })

    expect(onChange).not.toHaveBeenCalled()

    dispatchTransaction(command)

    expect(onChange).toHaveBeenCalled()
  })

  it("calls onChange with the updated state", async () => {
    const node = document.createElement("div")
    const onChange = jest.fn()
    const command = (
      state: EditorState,
      dispatch?: (tx: Transaction) => void
    ) => {
      if (!dispatch) return false
      dispatch(state.tr.insertText("Prepended content. "))
      return true
    }

    const { dispatchTransaction } = createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
    })

    dispatchTransaction(command)

    const updatedEditorState = onChange.mock.calls[0][0] as EditorState

    expect(updatedEditorState.toJSON()).toMatchSnapshot()
  })

  it("defaults to editable", async () => {
    const node = document.createElement("div")
    const onChange = jest.fn()

    createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
    })

    expect(node.children[0]?.getAttribute("contenteditable")).toBe("true")
  })

  it("respects initial isEditable value", async () => {
    const node = document.createElement("div")
    const onChange = jest.fn()

    createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
      isEditable: () => false,
    })

    expect(node.children[0]?.getAttribute("contenteditable")).toBe("false")
  })

  it("updates editable status", async () => {
    let editable = true
    const node = document.createElement("div")
    const onChange = jest.fn()
    const noopCommand = (
      state: EditorState,
      dispatch?: (tx: Transaction) => void
    ) => {
      if (!dispatch) return false
      dispatch(state.tr)
      return true
    }

    const { dispatchTransaction } = createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
      isEditable: () => editable,
    })

    expect(node.children[0]?.getAttribute("contenteditable")).toBe("true")

    editable = false
    dispatchTransaction(noopCommand)

    expect(node.children[0]?.getAttribute("contenteditable")).toBe("false")
  })

  it("aria-labelledby is present", async () => {
    const editable = true
    const node = document.createElement("div")
    const onChange = jest.fn()
    const noopCommand = (
      state: EditorState,
      dispatch?: (tx: Transaction) => void
    ) => {
      if (!dispatch) return false
      dispatch(state.tr)
      return true
    }

    const { dispatchTransaction } = createRichTextEditor({
      node,
      onChange,
      attributes,
      initialEditorState: testEditorState,
      isEditable: () => editable,
    })

    dispatchTransaction(noopCommand)

    expect(node.children[0]?.getAttribute("aria-labelledby")).toBe(
      "label-text-123"
    )
  })
})
