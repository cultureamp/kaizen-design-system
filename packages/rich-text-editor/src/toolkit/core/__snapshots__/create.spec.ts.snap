// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRichTextEditor calls onChange with the updated state 1`] = `
{
  "doc": {
    "content": [
      {
        "content": [
          {
            "text": "Prepended content. Example content",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "doc",
  },
  "selection": {
    "anchor": 20,
    "head": 20,
    "type": "text",
  },
}
`;
